### PR TITLE
Improve API specification

### PR DESCRIPTION
**Pull request by erikgb**
The proposed changes are extracted from https://github.com/cert-manager/trust-manager/pull/475. Before introducing CEL in the API specification, which will lift the K8s version requirements, I propose adding more "simple" validations to the API spec. Also adding processing instructions to API to ensure it will work smoothly for users using SSA.

### CODE REVIEW

File: docs/api/api.md [hunk 1]
```diff
@@ -365,13 +367,16 @@ func (in *BundleSource) DeepCopyInto(out *BundleSource)
...
+## type [BundleSpec](<https://github.com/cert-manager/trust-manager/blob/main/pkg/apis/trust/v1alpha1/types_bundle.go#L60-L69>)
 
 BundleSpec defines the desired state of a Bundle.
 
 ```go
 type BundleSpec struct {
     // Sources is a set of references to data whose data will sync to the target.
+    // +listType=atomic
+    // +kubebuilder:validation:MinItems=1
+    // +kubebuilder:validation:MaxItems=100
```
Comment (erikgb) [inline docs/api/api.md:379]: MaxItems is added to pass the cost requirements when CEL validations are added to the API.

File: pkg/apis/trust/v1alpha1/types_bundle.go [hunk 2]
```diff
@@ -59,6 +59,9 @@ type BundleList struct {
 // BundleSpec defines the desired state of a Bundle.
 type BundleSpec struct {
 	// Sources is a set of references to data whose data will sync to the target.
+	// +listType=atomic
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=100
```
Comment (erikgb) [inline pkg/apis/trust/v1alpha1/types_bundle.go:64]: MaxItems is added to pass the cost requirements when CEL validations are added to the API.

File: pkg/apis/trust/v1alpha1/types_bundle.go [hunk 3]
```diff
@@ -67,6 +70,7 @@ type BundleSpec struct {
 
 // BundleSource is the set of sources whose data will be appended and synced to
 // the BundleTarget in all Namespaces.
+// +structType=atomic
```
Comment (inteon) [inline pkg/apis/trust/v1alpha1/types_bundle.go:73]: 👍

File: pkg/apis/trust/v1alpha1/types_bundle.go [hunk 4]
```diff
@@ -96,14 +100,17 @@ type BundleSource struct {
 
 // BundleTarget is the target resource that the Bundle will sync all source
 // data to.
+// +structType=atomic
```
Comment (inteon) [inline pkg/apis/trust/v1alpha1/types_bundle.go:103]: 👍

Comment (inteon) [inline pkg/apis/trust/v1alpha1/types_bundle.go:103]: @erikgb I don't think this has to be atomic, right?

File: pkg/apis/trust/v1alpha1/types_bundle.go [hunk 5]
```diff
@@ -117,16 +124,21 @@ type BundleTarget struct {
 }
 
 // AdditionalFormats specifies any additional formats to write to the target
+// +structType=atomic
```
Comment (inteon) [inline pkg/apis/trust/v1alpha1/types_bundle.go:127]: @erikgb same here, does not have to be atomic, right?

File: pkg/apis/trust/v1alpha1/types_bundle.go (hunk 4 above)
Comment (erikgb) [reply pkg/apis/trust/v1alpha1/types_bundle.go:103]: Agree, will fix.

File: pkg/apis/trust/v1alpha1/types_bundle.go [hunk 6]
```diff
@@ -171,6 +187,7 @@ type SourceObjectKeySelector struct {
 
 	// Key of the entry in the object's `data` field to be used.
 	//+optional
+	// +kubebuilder:validation:MinLength=1
```
Comment (inteon) [inline pkg/apis/trust/v1alpha1/types_bundle.go:190]: @erikgb what does the combination `optional` and `MinLength=1` mean?
Do you have examples of yaml that works and does not work with this configuration + how it is different from eg. just `optional`?

Comment (erikgb) [reply pkg/apis/trust/v1alpha1/types_bundle.go:190]: Just `optional` will allow the empty string. Which doesn't make sense here.

File: pkg/apis/trust/v1alpha1/types_bundle.go (hunk 4 above)
Comment (erikgb) [reply pkg/apis/trust/v1alpha1/types_bundle.go]: Fixed

File: pkg/apis/trust/v1alpha1/types_bundle.go (hunk 5 above)
Comment (erikgb) [reply pkg/apis/trust/v1alpha1/types_bundle.go]: Fixed

**Comment by inteon** [comment]
Great work, this is something that could be useful in our other APIs too.

/approve
/lgtm

**Comment by erikgb** [comment]
> Great work, this is something that could be useful in our other APIs too.

Yeah, I am going on a "round around the block". But I need static CRDs in Git first. 😉

**Pull request merged**